test/unit/asrouter/ASRouter.test.js

```diff
@@ -3048,4 +3132,68 @@ describe("ASRouter", () => {
       assert.equal(result, "bar");
     });
   });
+
+  describe("#_onLocaleChanged", () => {
+    it("should call _maybeUpdateL10nAttachment in the handler", async () => {
+      await createRouterAndInit();
+      sandbox.stub(Router, "_maybeUpdateL10nAttachment").resolves(null);
+      await Router._onLocaleChanged();
+
+      assert.calledOnce(Router._maybeUpdateL10nAttachment);
+    });
+  });
+
+  describe("#_maybeUpdateL10nAttachment", () => {
+    it("should update the l10n attachment if the locale was changed", async () => {
+      const getter = sandbox.stub();
+      getter.onFirstCall().returns("en-US");
+      getter.onSecondCall().returns("fr");
+      sandbox.stub(global.Services.locale, "appLocaleAsLangTag").get(getter);
+      const provider = {
+        id: "cfr",
+        enabled: true,
+        type: "remote-settings",
+        bucket: "cfr",
+      };
+      await createRouterAndInit([provider]);
+      sandbox.spy(Router, "setState");
+      sandbox.stub(Router, "loadMessagesFromAllProviders").resolves(null);
```

Comment:
nit: I would stub/spy on the methods at the beginning of the test and use `resetHistory` instead.

Reply:
I agree with you that stubbing/spying should be done in the beginning in general, though I think the ordering does matter in this case. Because it had to call `createRouterAndInit()` with a special provider (the default fake provider for Remote Settings doesn't have `id` as `cfr` or `cfr-fxa`, so it won't trigger the attachment downloading work), spying `Router` prior to `createRouteAndInit()` won't work as it will mutate `Router`.

I did simplify this test a bit though.